### DISCUSSION TITLE
Parallel admin operations and misc notes

### DISCUSSION THREAD

**Discussion by terribleplan** [seaweedfs/seaweedfs#3917] [General]
I have been working on recovering data from a (very) messed up cluster and moving it to an alternative storage solution. The biggest headache I have had is the inability to do certain things in parallel (as well as ability to do certain things at all). I have a bunch of storage nodes that only have 1Gb connectivity, I also have a management node that has limited disk space but 2x10Gb connectivity. Some bottlenecks with my recovery process/things I would like to see:

1. Have the ec shards copied from their remote location in parallel.
1. Tell `ec.rebuild` which server it should do the rebuild on. No single server has enough shards on it to really be worth doing the rebuild on, but rebuilding on the server with the fast link would help (assuming it could copy remote shards in parallel).
1. Have `volume.fix.replication` parallelize itself (at least when it can with different source/destination volume servers)

There are a few other general issues I am having, that are probably much bigger/harder to tackle:
1. Can't run _anything_ needing a `lock` in parallel. It has been quite annoying being unable to do most other management activity on my cluster when `ec.rebuild` is sitting there copying a shard from one system to another. Having locking be done smartly at the collection, volume, node, etc. layer would be huge. Certainly some things might need a full "master" lock, but it seems overkill to hold the only lock the entire time. Doing multiple `volumeServer.evacuate` in parallel, or I running something like `volume.deleteEmpty` while rebuilding ec are a few specific things I have wanted to do.
1. Critical and long-running commands stopping after running into their first error. As I said this cluster is _very_ messed up, so at the moment there are ec volumes that cannot be recovered. It'd be really cool if things like `ec.rebuild` didn't fail at the first issue, but instead did all the rebuilding it could and reported the unrecoverable volumes at the end. I believe `volume.fix.replication` has a similar issue if there are no copies of a volume (or if a volume server exits/segfaults in the middle of some operation, or if it fails to update cluster topology). It is really frustrating to leave something like this to run overnight and come back to it having stopped part way due to those issues. (I think even `weed fix` has this issue in some cases)
1. Segfaults/panics in volume servers. Ideally the volume server shouldn't die due to bad idx/sdx files or when the dat file is corrupted/empty/whatever. Auto-recovery (like what `weed fix` does), or at least ignoring bad volumes would be cool. Also it'd be neat if `weed fix deleted` and recreated bad idx/sdx files itself.
1. Better progress indication. Some sort of overall volume/collection progress indicator for ec rebuilds, replication fix, balancing ec, etc. would be nice. In an ec rebuild there is no indication of how much of each EC shard has been copied or how far through the RS check/rebuild you are.  Also would be nice to see a percentage anywhere you see a byte count (or at least a total bytes number somewhere).

I have been working on recovering my data for almost two weeks now due to the above issues and dozens of TB of data involved. The initial issue was my fault (expecting the nodes to be reliable on a non-journaled filesystem), but it didn't take too terribly long to run fsck, convert to ext4, etc. What took a long time was getting things to a point where volume servers would properly start, and it is taking longer still to get to a point where the data in the cluster is actually readable. I am also unsure at this point if all of the data will be recoverable or if that data is actually correct.

Before the full scope of the filesystem/hardware issues became clear I was having other issues that I can't really properly/reliably report (running out of volumes very quickly, lots of unused/unusable empty or mostly-empty volumes, _negative_ "free" volumes on master, exceedingly slow speeds for read and write) which may or may not have been caused by the bad node(s)/disk(s).

This basically serves as my reflections on my overall experience with seaweed (which was a positive one when everything was working properly), as well as some notes about where things fell apart or issues I had. Not sure I would have had better luck with alternatives, but they may have a better recovery story since they've had more time to encounter and address this sort of thing or may not have allowed it to happen at all. I wish this project the best of luck as its development continues, and I may check it out again later or if I start feeling pain from a too-many small files situation.

**Comment by chrislusf** [comment]
Thanks for your precious feedback! Will study each item and see how to improve.

"running out of volumes very quickly, lots of unused/unusable empty or mostly-empty volumes" this actually has just been addressed yesterday.

**Comment by kmlebedev** [comment]
> Segfaults/panics in volume servers. Ideally the volume server shouldn't die due to bad idx/sdx files

https://github.com/seaweedfs/seaweedfs/issues/2945
